crates/types/src/nodes_config.rs

```diff
@@ -57,15 +60,84 @@ pub enum Role {
 #[display(fmt = "v{}", _0)]
 pub struct ConfigVersion(u32);
 
+impl ConfigVersion {
+    pub const INVALID: ConfigVersion = ConfigVersion(0);
+    pub const MIN: ConfigVersion = ConfigVersion(1);
+}
+
+impl Default for ConfigVersion {
+    fn default() -> Self {
+        Self::MIN
+    }
+}
+
 #[derive(Debug, Clone, Eq, PartialEq, Default)]
 #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
-struct NodesConfiguration {
+pub struct NodesConfiguration {
     version: ConfigVersion,
     nodes: HashMap<PlainNodeId, MaybeNode>,
     controllers: Vec<ControllerConfig>,
     name_lookup: HashMap<String, PlainNodeId>,
 }
 
+impl NodesConfiguration {
+    /// The currently loaded nodes configuration. This returns None if no configuration
+    /// is loaded yet, this is will be only the case before attaching to the controller
+    /// node on startup.
+    ///
+    /// After attaching, it's safe to unwrap this, or use current_unchecked() variant.
+    pub fn current() -> Option<Arc<NodesConfiguration>> {
+        CURRENT_CONFIG.load_full()
+    }
+
+    pub fn current_unchecked() -> Arc<NodesConfiguration> {
+        CURRENT_CONFIG.load_full().unwrap()
+    }
+
+    /// The currently loaded configuration version or ConfigVersion::INVALID if
+    /// no configuration is loaded yet.
+    pub fn current_version() -> ConfigVersion {
+        CURRENT_CONFIG
+            .load()
+            .as_ref()
+            .map(|config| config.version)
+            .unwrap_or(ConfigVersion::INVALID)
+    }
+
+    /// Silently ignore the config if it's older than existing
+    pub fn set_as_current_if_newer(self) -> ConfigVersion {
+        let current = CURRENT_CONFIG.load_full();
+        if current.is_none() {
+            let after_update = CURRENT_CONFIG.compare_and_swap(&current, Some(Arc::new(self)));
+            return after_update
+                .as_ref()
+                .map(|c| c.version)
+                .unwrap_or(ConfigVersion::INVALID);
+        }
+
+        let current = current.unwrap();
+        if self.version > current.version {
+            let after_update = CURRENT_CONFIG.compare_and_swap(&current, Some(Arc::new(self)));
+            after_update
+                .as_ref()
+                .map(|c| c.version)
+                .unwrap_or(ConfigVersion::INVALID)
+        } else {
+            current.version
+        }
+    }
```

Comment:
I think this method might drop some node configs silently in case of concurrent accesses. I think by introducing a loop, we could fix it:

```
pub fn set_as_current_if_newer(self) -> ConfigVersion {
    let self_version = self.version;
    let arc_self = Arc::new(self);
    loop {
        let current = CURRENT_CONFIG.load_full();
        let current_version = current
            .as_ref()
            .map(|c| c.version())
            .unwrap_or(ConfigVersion::INVALID);

        if self_version > current_version {
            let after_update =
                CURRENT_CONFIG.compare_and_swap(&current, Some(Arc::clone(&arc_self)));

            match (current, after_update.as_ref()) {
                (None, None) => return self_version,
                (Some(current), Some(after)) if Arc::ptr_eq(&current, after) => {
                    return self_version;
                }
                (_, _) => {}
            }
        } else {
            return current_version;
        }
    }
}
```

Reply:
I'll explain in a comment. This is not meant to be called concurrently, a single node-level task will be responsible for pushing updates.